### ISSUE TITLE
A weird bug when using ByteData

### ISSUE THREAD

**Issue by 623637646** [dart-lang/sdk#48924]
# Example code:
```dart
import 'dart:typed_data';

void main() {
  callAFunction();

  for (int v = -5000; v <= -32; v++) {
    if (v == -32) {
      final _buf = Uint8List(64);
      final _d = ByteData.view(_buf.buffer);
      final thisIsUnused = Int8List.view(_buf.buffer);
      _d.setInt8(0, v);

      final u = Unpacker(_buf);
      final r = u.unpackInt();

      print('result $r');
    } else {
      final _buf = Uint8List(64);
      var _d = ByteData.view(_buf.buffer, _buf.offsetInBytes);
      _d.setInt8(0, -46);

      final u = Unpacker(_buf);
      final r = u.unpackInt();
    }
  }
}

void callAFunction() {
  final bytes = Uint8List.fromList([0]);
  final d = ByteData.view(bytes.buffer, 0);
  final i = d.getInt8(0);
}

class Unpacker {
  Unpacker(Uint8List _list)
      : _d = ByteData.view(_list.buffer, _list.offsetInBytes);

  final ByteData _d;

  int? unpackInt() {
    final b = _d.getUint8(0);

    int? v;
    if (b <= 0x7f || b >= 0xe0) {
      v = _d.getInt8(0);
    } else if (b == 0xd0) {
      v = _d.getInt8(0);
    } else if (b == 0xd1) {
      v = _d.getInt16(0);
    } else if (b == 0xd3) {
      v = _d.getInt64(0);
    } else if (b == 0xc0) {
      v = null;
    }
    return v;
  }
}


```
# Issues
Run this code, it will print `result 224`. But if you comment out the `callAFunction();` in `main`. It will print `result -32`.

By right, the calling `callAFunction();` or not shouldn't affect the result. 

# Pay attention
![#f03c15](https://via.placeholder.com/15/f03c15/000000?text=+) **I can only reproduce this bug on a dart project. Can't reproduce it on a flutter project.**

This is my `pubspec.yaml`.
```yaml
name: mytest
description: A new Flutter project.
version: 1.0.0+1

environment:
  sdk: ">=2.16.2 <3.0.0"

dev_dependencies:
  test: ^1.16.0

```

If I add this to the `pubspec.yaml`, The project will become a flutter project from a dart project. Then the bug will be gone.
```yaml
dev_dependencies:
  flutter_test:
    sdk: flutter
```
# Other weird cases

And if you do those things, The result will also be affected.

1. Change `int v = -5000` to `int v = -3000`, Will print`result -32`.
2. Comment out the unused code `final thisIsUnused = Int8List.view(_buf.buffer);`, In run mode, it will print `result -32`.  In debug mode, sometimes it prints `result -32`, sometimes it's `result 224`.
3. Comment out this code, Will print`result -32`.

```dart
final _buf = Uint8List(64);
var _d = ByteData.view(_buf.buffer, _buf.offsetInBytes);
_d.setInt8(0, -46);

final u = Unpacker(_buf);
final r = u.unpackInt();
```

4. Comment out this code, Will print`result -32`.
```dart
else if (b == 0xc0) {
  v = null;
}
```

# Demo to reproduce this bug

https://github.com/623637646/dart_vm_bug_demo

# My environment
* Dart SDK Version: Dart SDK version: 2.16.2 (stable) (Tue Mar 22 13:15:13 2022 +0100) on "macos_x64"
* MacOSX

**Comment by asashour** [comment]
It seems to be fixed, at least with `Dart SDK version: 2.18.0-66.0.dev (dev) (Wed Apr 27 13:10:04 2022 -0700) on "windows_x64"`

**Comment by 623637646** [comment]
Thanks for the reply @asashour . I use `2.17.0-266.7.beta (beta) (Mon Apr 25 14:54:53 2022 +0000) on "macos_arm64"` to verify. It's fixed.
